### PR TITLE
Turn on Parallel Testing for UI Tests

### PR DESCRIPTION
**Pull request by jostnes**
### Description
~(This branch was created based on [`refactor-to-turn-on-parallel-testing`](https://github.com/wordpress-mobile/WordPress-iOS/pull/20655), once that's merged there should be less changes on this PR) - PR merged~

This PR turns on parallel testing for UI tests (JPiOS), running on 2 simulators. There is one issue I couldn't figure out, so I am putting this here for discussion if we want to continue with this or not. 

For reasons I still don't understand, the first test on the first runner will always fail on the first try but passes on the retry (the same behavior that @tiagomar noticed a few months back). Here's an example when the first test suite is `StatsTest`:
image

I tried disabling that test to see if it was test related and the results were worse, both "first" tests on the runners failed:
image

And the failures all happen at the same place, after tapping on the Continue button after entering the password to log in, this is the screenshot from the failures:
image

You can see that the spinner icon is still displayed on the Continue button, so it doesn't continue to show the next screen. I tried updating the timeout to wait as long as 90 seconds but the results were the same - it would still fail on the first run. 

### Example Results
Before change - ~23m 23s ([example](https://buildkite.com/automattic/wordpress-ios/builds/14117#018803ad-5ecc-4ca0-ace1-39f0e5ed95b3))
After change (including that 1 consistent failure) - ~14m 35s ([example](https://buildkite.com/automattic/wordpress-ios/builds/14116#01880393-2e53-4248-8772-be9a8901d0d7))

Even with the consistent failing test, the time taken to complete the test run is already much faster. Is this a trade-off we're willing to take? I think we should, but open to feedback and thoughts from the team. 

### Testing
CI should be green, and tests should run in parallel on 2 simulators. To validate, check (should be generated in the UI test steps) the test plan log and ensure that they are two runners:
image

### CODE REVIEW

**Comment by pachlava** [comment]
> (This branch was created based on https://github.com/wordpress-mobile/WordPress-iOS/pull/20655, once that's merged there should be less changes on this PR)

I believe changing the base of current branch from `trunk` to `refactor-to-turn-on-parallel-testing` would automatically solve the issue, and current PR will contain one commit only.

**Comment by jostnes** [comment]
> I believe changing the base of current branch from trunk to refactor-to-turn-on-parallel-testing would automatically solve the issue, and current PR will contain one commit only.

I did it the other way around, after merging `refactor-to-turn-on-parallel-testing` to trunk (as I wanted the changes to be on trunk first), I updated this PR to latest `trunk` and the commits from that branch are now removed from this PR. And with that done, this is now ready for review!

**Comment by tiagomar** [comment]
It looks like, by turning Parallel Testing ON, the generated `report.junit` file is now empty.
```
<?xml version='1.0' encoding='UTF-8'?>
<testsuites tests='0' failures='0'/>
```
This will probably impact Buildkite's Test Analytics and also our plans to properly report UI Tests failures on Slack.

The "good" part is that apparently we are not alone:
https://github.com/xcpretty/xcpretty/issues/344
https://github.com/fastlane/fastlane/issues/20809

**Comment by mokagio** [comment]
@jostnes 

> Even with the consistent failing test, the time taken to complete the test run is already much faster. Is this a trade-off we're willing to take? I think we should, but open to feedback and thoughts from the team.

+1. If I understand correctly, the first test on the first Simulator fails on the first try, but passes on the retry. That makes the impact negligible. I think it would be great to get the speed improvement ASAP and then focus on finding the reason for that behavior.

@tiagomar 

> It looks like, by turning Parallel Testing ON, the generated `report.junit` file is now empty.
> [...]
> This will probably impact Buildkite's Test Analytics and also our plans to properly report UI Tests failures on Slack.

Test Analytics reports results directly from the XCTest runner, so it should not be affected by what happens with the generated report files. 😮‍💨

One of the [comments](https://github.com/fastlane/fastlane/issues/20809#issuecomment-1542451442) in the Fastlane issue you linked suggests having a look at a dedicated tool for parsing Xcode's `xcresult` files, https://github.com/a7ex/xcresultparser. I had never heard of that tool but I experimented with [`xcparse`](https://github.com/ChargePoint/xcparse) in the past. Regardless of the tool, I think the idea of rolling our own JUnit report generation is one worth keeping in mind. I think we might need to stick with JUnit for the report format because that is what the [Buildkite annotation logic expects](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/blob/trunk/bin/annotate_test_failures). Or, we could switch to a different format and write a new annotation command that support it.

**Comment by jostnes** [comment]
> If I understand correctly, the first test on the first Simulator fails on the first try, but passes on the retry. That makes the impact negligible.

@mokagio yup that is the current behavior. and from what @tiagomar discussed, it also isn't a new behavior (though it's not happening currently in latest trunk when i last checked)

@tiagomar @mokagio for the JUnit file, one thing i noticed is that on WCiOS (parallel testing has been turned on for WCiOS), the test output is an XML file ([example build](https://buildkite.com/automattic/woocommerce-ios/builds/13142#0188061d-b27a-4dd2-a72f-322c7e7ece6d)) instead of JUnit (the content of the file is similar though). i've not pinpointed where this is done in the code but perhaps if we update the results output from JUnit to XML we could solve this issue?

**Comment by mokagio** [comment]
@jostnes great observation about WooCommerce iOS 👏 

The lines responsible for generating the report are [these](https://github.com/woocommerce/woocommerce-ios/blob/157350788a96d0539e8162a4130f5d5ea8b8deb5/fastlane/Fastfile#L1145-L1151).

```ruby
    result_bundle: true,
    # settins for XML test report generation via `trainer`
    output_types: '',
    fail_build: false
  )

  trainer(path: lane_context[SharedValues::SCAN_GENERATED_XCRESULT_PATH], fail_build: true)
```

Basically, Woo bypasses the build-in report generation that Fastlane's `run_test` offers, and uses `trainer` instead. [The `trainer` docs state it can generate JUnit reports, too](https://docs.fastlane.tools/actions/trainer/). We might be in luck 🤞

**Comment by jostnes** [comment]
I'm converting this PR back to draft temporarily to test out using `trainer` and see if it can fix the empty output issue (testing done in a [separate branch](https://github.com/wordpress-mobile/WordPress-iOS/pull/20646/files))

**Comment by jostnes** [comment]
Looks like using trainer works! Thanks for pointing to where the change is made @mokagio! Change made in https://github.com/wordpress-mobile/WordPress-iOS/pull/20656/commits/87a8dcd6e696ca7bcfb905e0391689187cca4bde However, I decided to leave it to use the default extension (XML instead of JUnit) to match the other repos (WCiOS, WCA, WPA/JPA all return results in XML file). The file content of the XML file is the same as the JUnit, so the annotation when a fail happens still works, see: [build](https://buildkite.com/automattic/wordpress-ios/builds/14161).

This PR is now ready for review again (for the first test in the first runner failing issue, I'll address that separately as that might take some time to figure out)

**Review by tiagomar — APPROVED** [review]
>  If I understand correctly, the first test on the first Simulator fails on the first try, but passes on the retry.

@mokagio, it's missing just one detail: it actually happens for each Simulator, not just the first one. So, first run of the first test fails in each Simulator. However, this was definitely not introduced with this PR. It actually reproduced locally when the Simulator's clones ran for the first time, so it's not something that only happens in CI (🎉). I'll spend some more time on it next week.

@jostnes, the parallelization works well on CI and locally and I'm glad to see the report issue figured out. 🎉 :shipit:

**Pull request merged**
